executor/distsql.go

```diff
@@ -586,99 +553,124 @@ func (e *IndexLookUpExecutor) startIndexWorker(goCtx goctx.Context, kvRanges []k
+		task := w.buildTableTask(handles)
+		select {
+		case <-goCtx.Done():
+			return
+		case <-w.finished:
+			return
+		case w.workCh <- task:
+			w.resultCh <- task
 		}
 	}
 }
 
-func (worker *indexWorker) close() {
-	worker.wg.Wait()
-}
-
 // tableWorker is used by IndexLookUpExecutor to maintain table lookup background goroutines.
 type tableWorker struct {
-	wg sync.WaitGroup
+	workCh         <-chan *lookupTableTask
+	finished       <-chan struct{}
+	buildTblReader func(ctx goctx.Context, handles []int64) (Executor, error)
+	keepOrder      bool
+	handleIdx      int
+}
+
+func (e *IndexLookUpExecutor) buildTableReader(goCtx goctx.Context, handles []int64) (Executor, error) {
+	tableReader, err := e.dataReaderBuilder.buildTableReaderFromHandles(goCtx, &TableReaderExecutor{
+		baseExecutor: newBaseExecutor(e.schema, e.ctx),
+		table:        e.table,
+		tableID:      e.tableID,
+		dagPB:        e.tableRequest,
+	}, handles)
+	if err != nil {
+		log.Error(err)
+		return nil, errors.Trace(err)
+	}
+	return tableReader, nil
 }
 
 // startTableWorker launch some background goroutines which pick tasks from workCh and execute the task.
-func (e *IndexLookUpExecutor) startTableWorker(goCtx goctx.Context, workCh <-chan *lookupTableTask, finished <-chan struct{}) {
-	worker := &e.tableWorker
+func (e *IndexLookUpExecutor) startTableWorker(goCtx goctx.Context, workCh <-chan *lookupTableTask) {
 	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
-	worker.wg.Add(lookupConcurrencyLimit)
+	e.tblWorkerWg.Add(lookupConcurrencyLimit)
 	for i := 0; i < lookupConcurrencyLimit; i++ {
+		worker := &tableWorker{
+			workCh:         workCh,
+			finished:       e.finished,
+			buildTblReader: e.buildTableReader,
+			keepOrder:      e.keepOrder,
+			handleIdx:      e.handleIdx,
+		}
 		goCtx1, cancel := goctx.WithCancel(goCtx)
 		go func() {
-			worker.pickAndExecTask(e, workCh, goCtx1, finished)
+			worker.pickAndExecTask(goCtx1)
 			cancel()
-			worker.wg.Done()
+			e.tblWorkerWg.Done()
 		}()
 	}
 }
 
 // pickAndExecTask picks tasks from workCh, and execute them.
-func (worker *tableWorker) pickAndExecTask(e *IndexLookUpExecutor, workCh <-chan *lookupTableTask, goCtx goctx.Context, finished <-chan struct{}) {
+func (w *tableWorker) pickAndExecTask(goCtx goctx.Context) {
 	for {
 		// Don't check ctx.Done() on purpose. If background worker get the signal and all
 		// exit immediately, session's goroutine doesn't know this and still calling Next(),
 		// it may block reading task.doneCh forever.
 		select {
-		case task, ok := <-workCh:
+		case task, ok := <-w.workCh:
 			if !ok {
 				return
 			}
-			e.executeTask(task, goCtx)
-		case <-finished:
+			w.executeTask(goCtx, task)
+		case <-w.finished:
 			return
 		}
 	}
 }
 
-func (worker *tableWorker) close() {
-	worker.wg.Wait()
-}
-
 // Open implements the Executor Open interface.
 func (e *IndexLookUpExecutor) Open(goCtx goctx.Context) error {
 	kvRanges, err := e.indexRangesToKVRanges()
```

Comment:
1. can we remove `e.indexRangesToKVRanges` and use `indexRangesToKVRanges` directly ?
2. can we merge `e.Open` and `e.open` to `e.Open` ?

Reply:
2. No, the open method is used by dataReaderBuilder with a datum range.